pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentRestletResource.java

```diff
@@ -19,255 +19,409 @@
-    segmentName = URIUtils.decode(segmentName);
-    // segmentName will never be null,otherwise we would reach the method toggleStateOrListMetadataForAllSegments()
-    CommonConstants.Helix.TableType tableType = Constants.validateTableType(tableTypeStr);
-    StateType stateType = Constants.validateState(stateStr);
-    if (stateStr == null) {
-      // This is a list metadata operation
-      return listSegmentMetadataInternal(tableName, segmentName, tableType);
-    } else {
-      // We need to toggle state
-      if (tableType == null) {
-        List<String> realtimeSegments =
-            _pinotHelixResourceManager.getSegmentsFor(TableNameBuilder.REALTIME.tableNameWithType(tableName));
-        List<String> offlineSegments =
-            _pinotHelixResourceManager.getSegmentsFor(TableNameBuilder.OFFLINE.tableNameWithType(tableName));
-        if (realtimeSegments.contains(segmentName)) {
-          tableType = CommonConstants.Helix.TableType.REALTIME;
-        } else if (offlineSegments.contains(segmentName)) {
-          tableType = CommonConstants.Helix.TableType.OFFLINE;
-        }
-      }
-      return toggleStateInternal(tableName, stateType, tableType, segmentName, _pinotHelixResourceManager);
+      @ApiParam(value = "MUST be null") @QueryParam("state") String stateStr,
+      @ApiParam(value = "OFFLINE|REALTIME") @QueryParam("type") String tableTypeStr)
+      throws JsonProcessingException {
+    if (stateStr != null) {
+      throw new WebApplicationException("Cannot toggle segment state", Status.FORBIDDEN);
+    }
+
+    List<String> tableNamesWithType =
+        getExistingTableNamesWithType(tableName, Constants.validateTableType(tableTypeStr));
+    List<Map<String, String>> resultList = new ArrayList<>(tableNamesWithType.size());
+    for (String tableNameWithType : tableNamesWithType) {
+      // NOTE: DO NOT change the format for backward-compatibility
+      Map<String, String> resultForTable = new LinkedHashMap<>();
+      resultForTable.put("tableName", tableNameWithType);
+      resultForTable.put("segments",
+          JsonUtils.objectToString(_pinotHelixResourceManager.getServerToSegmentsMap(tableNameWithType)));
+      resultList.add(resultForTable);
     }
+    return resultList;
   }
 
+  @Deprecated
   @GET
   @Path("tables/{tableName}/segments/metadata")
   @Produces(MediaType.APPLICATION_JSON)
-  @ApiOperation(value = "Lists metadata all segments of table", notes = "Lists segment metadata")
-  // TODO: 1. stateStr is not used
-  // TODO: 2. this method returns the map from server to segments instead of segment metadata
-  public String listMetadataForAllSegments(
+  @ApiOperation(value = "Get a map from server to segments hosted by the server (deprecated, use 'GET /segments/{tableName}/servers' instead)", notes = "Get a map from server to segments hosted by the server (deprecated, use 'GET /segments/{tableName}/servers' instead)")
+  public List<Map<String, String>> getServerToSegmentsMapDeprecated2(
```

Comment:
This API currently returns a String. please keep the return type the same, we can change the types in the new APIs that you have added.

Reply:
The string for the existing API is a JSON Object string for `List<Map<String, String>>`, which is the same as the current API (the actual return will also be a JSON Object string).